### PR TITLE
plan: calculate stats profile in logical plan.

### PR DESCRIPTION
**Pull request by hanfei1991**
I implement the statsProfile propagation for selection, projection, aggregation, topn, limit and datasource.
We can use the distinct value for every operator in near future.

### CODE REVIEW

File: plan/stats.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+func (p *Projection) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	p.profile = &statsProfile{
+		count:       childProfile.count,
+		cardinality: make([]float64, len(childProfile.cardinality)),
```
Comment (coocood) [inline plan/stats.go]: The length should be `p.schema.Len()`?

File: plan/stats.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	var gbyCols []*expression.Column
+	for _, gbyExpr := range p.GroupByItems {
+		cols := expression.ExtractColumns(gbyExpr)
+		gbyCols = append(gbyCols, cols...)
+	}
+	count := getCardinality(gbyCols, p.children[0].Schema(), childProfile)
+	p.profile = &statsProfile{
+		count:       count,
+		cardinality: make([]float64, len(childProfile.cardinality)),
```
Comment (coocood) [inline plan/stats.go]: Ditto

File: plan/stats.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i]
+		if p.profile.cardinality[i] > p.profile.count {
+			p.profile.cardinality[i] = p.profile.count
+		}
+	}
+	return p.profile
+}
+
+func getCardinality(cols []*expression.Column, schema *expression.Schema, profile *statsProfile) float64 {
```
Comment (coocood) [inline plan/stats.go]: Add comments about the algorithm.
It seems like just get the max cardinality of all the columns, this doesn't apply to group by columns.

Comment (hanfei1991) [inline plan/stats.go]: I think it's ok to use this strategy to aggregation

File: plan/stats.go (hunk 1 above)
Comment (hanfei1991) [inline plan/stats.go]: p.schema.Len is equal to len(childProfile.cardinality)

File: plan/stats.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+		return p.profile
+	}
+	p.profile = &statsProfile{
+		cardinality: make([]float64, p.schema.Len()),
+	}
+	for _, child := range p.children {
+		childProfile := child.(LogicalPlan).statsProfile()
+		p.profile.count += childProfile.count
+		for i := range p.profile.cardinality {
+			p.profile.cardinality[i] += childProfile.cardinality[i]
```
Comment (hanfei1991) [inline plan/stats.go]: I have no idea

File: plan/stats.go (hunk 1 above)
Comment (coocood) [inline plan/stats.go]: for `select a+1, a+1 from t`
the projection schema length is 2, its child cardinality length is 1.

**Comment by coocood** [comment]
LGTM

File: plan/stats.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/expression"
+)
+
+type statsProfile struct {
```
Comment (shenli) [inline plan/stats.go]: Add comment for the struct. For example why we need this and when will it be used.

**Comment by shenli** [comment]
LGTM

**Pull request merged**